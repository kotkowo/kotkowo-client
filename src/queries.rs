pub mod get_cat;
