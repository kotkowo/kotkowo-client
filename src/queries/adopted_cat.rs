#![allow(non_snake_case)] // please don't abuse, it's for rustler's generated atoms

pub use crate::queries::cat_commons::*;
pub use crate::queries::commons::*;
use crate::schema;

#[derive(cynic::QueryVariables, Debug)]
pub struct AdoptedCatQueryVariables<'a> {
    pub cat: CatFiltersInput<'a>,
    pub pagination: PaginationArg,

    #[cynic(skip_serializing_if = "Option::is_none")]
    pub between: Option<Vec<Option<DateTime>>>,
    #[cynic(skip_serializing_if = "Option::is_none")]
    pub sort: Option<Vec<Option<String>>>,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Query", variables = "AdoptedCatQueryVariables")]
pub struct AdoptedCatQuery {
    #[arguments(filters: { adoption_date: { between: $between }, cat: $cat }, pagination: $pagination, sort: $sort)]
    pub adopted_cats: Option<AdoptedCatEntityResponseCollection>,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct AdoptedCatEntityResponseCollection {
    pub data: Vec<AdoptedCatEntity>,
    pub meta: ResponseCollectionMeta,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct AdoptedCatEntity {
    pub id: Option<cynic::Id>,
    pub attributes: Option<AdoptedCat>,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct AdoptedCat {
    #[cynic(rename = "adoption_date")]
    pub adoption_date: DateTime,
    pub cat: Option<CatEntityResponse>,
}
