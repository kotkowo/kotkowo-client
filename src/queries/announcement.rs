#![allow(non_snake_case)] // please don't abuse, it's for rustler's generated atoms

pub use crate::queries::commons::*;
use crate::schema;

#[derive(cynic::QueryVariables, Debug)]
pub struct ListAnnouncementsVariables<'a> {
    #[cynic(skip_serializing_if = "Option::is_none")]
    pub filters: Option<AnnouncementFiltersInput<'a>>,
    #[cynic(skip_serializing_if = "Option::is_none")]
    pub pagination: Option<PaginationArg>,
    #[cynic(skip_serializing_if = "Option::is_none")]
    pub sort: Option<Vec<Option<String>>>,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Query", variables = "ListAnnouncementsVariables")]
pub struct ListAnnouncements {
    #[arguments(filters: $filters, pagination: $pagination, sort: $sort)]
    pub announcements: Option<AnnouncementEntityResponseCollection>,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct AnnouncementEntityResponseCollection {
    pub data: Vec<AnnouncementEntity>,
    pub meta: ResponseCollectionMeta,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct AnnouncementEntity {
    pub attributes: Option<Announcement>,
    pub id: Option<cynic::Id>,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct Announcement {
    pub created_at: Option<DateTime>,
    pub updated_at: Option<DateTime>,
    pub title: String,
    pub image: UploadFileEntityResponse,
    pub published_at: Option<DateTime>,
}

#[derive(cynic::InputObject, Debug)]
pub struct AnnouncementFiltersInput<'a> {
    pub id: Option<IdfilterInput<'a>>,
    pub title: Option<StringFilterInput>,
    #[cynic(rename = "announcement_tags")]
    pub announcement_tags: Option<AnnouncementTagFiltersInput<'a>>,
    pub article: Option<ArticleFiltersInput<'a>>,
    pub created_at: Option<DateTimeFilterInput>,
    pub updated_at: Option<DateTimeFilterInput>,
    pub published_at: Option<DateTimeFilterInput>,
    pub and: Option<Vec<Option<AnnouncementFiltersInput<'a>>>>,
    pub or: Option<Vec<Option<AnnouncementFiltersInput<'a>>>>,
    pub not: Option<Box<AnnouncementFiltersInput<'a>>>,
}

#[derive(cynic::InputObject, Debug)]
pub struct ArticleFiltersInput<'a> {
    pub id: Option<IdfilterInput<'a>>,
    #[cynic(rename = "Introduction")]
    pub introduction: Option<StringFilterInput>,
    pub announcement: Option<Box<AnnouncementFiltersInput<'a>>>,
    pub content: Option<StringFilterInput>,
    pub created_at: Option<DateTimeFilterInput>,
    pub updated_at: Option<DateTimeFilterInput>,
    pub published_at: Option<DateTimeFilterInput>,
    pub and: Option<Vec<Option<ArticleFiltersInput<'a>>>>,
    pub or: Option<Vec<Option<ArticleFiltersInput<'a>>>>,
    pub not: Option<Box<ArticleFiltersInput<'a>>>,
}

#[derive(cynic::InputObject, Debug)]
pub struct AnnouncementTagFiltersInput<'a> {
    pub id: Option<IdfilterInput<'a>>,
    pub text: Option<StringFilterInput>,
    pub created_at: Option<DateTimeFilterInput>,
    pub updated_at: Option<DateTimeFilterInput>,
    pub and: Option<Vec<Option<AnnouncementTagFiltersInput<'a>>>>,
    pub or: Option<Vec<Option<AnnouncementTagFiltersInput<'a>>>>,
    pub not: Option<Box<AnnouncementTagFiltersInput<'a>>>,
}
