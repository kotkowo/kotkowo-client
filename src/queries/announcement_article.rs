#![allow(non_snake_case)] // please don't abuse, it's for rustler's generated atoms

pub use crate::queries::commons::*;
use crate::schema;

#[derive(cynic::QueryVariables, Debug)]
pub struct GetArticleVariables {
    pub id: cynic::Id,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Query", variables = "GetArticleVariables")]
pub struct GetArticle {
    #[arguments(id: $id)]
    pub announcement: Option<AnnouncementEntityResponse>,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct AnnouncementEntityResponse {
    pub data: Option<AnnouncementEntity>,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct AnnouncementEntity {
    pub attributes: Option<Announcement>,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct Announcement {
    pub article: Option<ArticleEntityResponse>,
    pub title: String,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct ArticleEntityResponse {
    pub data: Option<ArticleEntity>,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct ArticleEntity {
    pub attributes: Option<Article>,
    pub id: Option<cynic::Id>,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct Article {
    pub image: UploadFileEntityResponse,
    pub content: String,
    #[cynic(rename = "Introduction")]
    pub introduction: String,
}
