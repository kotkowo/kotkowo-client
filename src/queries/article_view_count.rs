#![allow(non_snake_case)] // please don't abuse, it's for rustler's generated atoms

pub use crate::queries::commons::*;
use crate::schema;

#[derive(cynic::QueryVariables, Debug)]
pub struct ArticleViewCountQueryVariables<'a> {
    pub article_id: Option<&'a cynic::Id>,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Query", variables = "ArticleViewCountQueryVariables")]
pub struct ArticleViewCountQuery {
    #[arguments(filters: { article: { id: { eq: $article_id } } })]
    pub article_views: Option<ArticleViewEntityResponseCollection>,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct ArticleViewEntityResponseCollection {
    pub meta: ResponseCollectionMeta,
}
