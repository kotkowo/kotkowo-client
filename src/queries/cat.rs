#![allow(non_snake_case)] // please don't abuse, it's for rustler's generated atoms

pub use crate::queries::commons::*;
use crate::schema;

pub use crate::queries::cat_commons::*;

#[derive(cynic::QueryVariables, Debug)]
pub struct GetCatVariables<'a> {
    pub id: &'a cynic::Id,
}

#[derive(cynic::QueryVariables, Debug)]
pub struct ListCatVariables<'a> {
    pub filters: CatFiltersInput<'a>,
    pub pagination: PaginationArg,

    #[cynic(skip_serializing_if = "Option::is_none")]
    pub sort: Option<Vec<Option<String>>>,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Query", variables = "ListCatVariables")]
pub struct ListCat {
    #[arguments(filters: $filters, pagination: $pagination, sort: $sort)]
    pub cats: Option<CatEntityResponseCollection>,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Query", variables = "GetCatVariables")]
pub struct GetCat {
    #[arguments(id: $id)]
    pub cat: Option<CatEntityResponse>,
}
