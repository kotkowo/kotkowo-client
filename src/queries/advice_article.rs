#![allow(non_snake_case)] // please don't abuse, it's for rustler's generated atoms

use crate::queries::{
    announcement::AnnouncementTagRelationResponseCollection,
    announcement_article::ArticleEntityResponse,
};
use crate::schema;

#[derive(cynic::QueryVariables, Debug)]
pub struct GetAdviceArticleVariables {
    pub id: cynic::Id,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Query", variables = "GetAdviceArticleVariables")]
pub struct GetAdviceArticle {
    #[arguments(id: $id)]
    pub advice: Option<AdviceEntityResponse>,
}
#[derive(cynic::QueryFragment, Debug)]
pub struct AdviceEntityResponse {
    pub data: Option<AdviceEntity>,
}
#[derive(cynic::QueryFragment, Debug)]
pub struct AdviceEntity {
    pub attributes: Option<Advice>,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct Advice {
    pub article: Option<ArticleEntityResponse>,
    pub title: String,
    pub tags: Option<AnnouncementTagRelationResponseCollection>,
}
