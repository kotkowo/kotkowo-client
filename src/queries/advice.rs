#![allow(non_snake_case)] // please don't abuse, it's for rustler's generated atoms

use crate::queries::announcement::AnnouncementTagRelationResponseCollection;
pub use crate::queries::commons::*;
use crate::schema;

use super::announcement::{AnnouncementTagFiltersInput, ArticleFiltersInput};

#[derive(cynic::QueryVariables, Debug)]
pub struct ListAdviceVariables<'a> {
    #[cynic(skip_serializing_if = "Option::is_none")]
    pub filters: Option<AdviceFiltersInput<'a>>,
    #[cynic(skip_serializing_if = "Option::is_none")]
    pub pagination: Option<PaginationArg>,
    #[cynic(skip_serializing_if = "Option::is_none")]
    pub sort: Option<Vec<Option<String>>>,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Query", variables = "ListAdviceVariables")]
pub struct ListAdvice {
    #[arguments(pagination: $pagination, filters: $filters, sort: $sort)]
    pub advices: Option<AdviceEntityResponseCollection>,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct AdviceEntityResponseCollection {
    pub data: Vec<AdviceEntity>,
    pub meta: ResponseCollectionMeta,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct AdviceEntity {
    pub attributes: Option<Advice>,
    pub id: Option<cynic::Id>,
}

#[derive(cynic::QueryFragment, Debug)]
pub struct Advice {
    pub title: String,
    pub image: UploadFileEntityResponse,
    pub tags: Option<AnnouncementTagRelationResponseCollection>,
}

#[derive(cynic::InputObject, Debug)]
pub struct AdviceFiltersInput<'a> {
    pub id: Option<IdfilterInput<'a>>,
    pub title: Option<StringFilterInput>,
    pub tags: Option<AnnouncementTagFiltersInput<'a>>,
    pub article: Option<ArticleFiltersInput<'a>>,
    pub created_at: Option<DateTimeFilterInput>,
    pub updated_at: Option<DateTimeFilterInput>,
    pub published_at: Option<DateTimeFilterInput>,
    pub and: Option<Vec<Option<AdviceFiltersInput<'a>>>>,
    pub or: Option<Vec<Option<AdviceFiltersInput<'a>>>>,
    pub not: Option<Box<AdviceFiltersInput<'a>>>,
}
