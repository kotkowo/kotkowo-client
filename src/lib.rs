#![feature(associated_type_defaults)]
mod errors;
mod models;
mod options;
mod queries;
mod schema;

pub use errors::*;
pub use models::{
    AdoptedCat, Advice, Age, Announcement, Article, Cat, Color, ExternalMedia, FoundCat,
    LookingForHomeCat, LostCat, Paged, Sex, Supporter,
};
pub use options::{
    AdviceFilter, AnnouncementFilter, BetweenDateTime, CatFilter, ExternalMediaFilter, Options,
};
pub use queries::commons::{ContactInformation, PaginationArg};

use queries::{
    cat::CatFiltersInput, commons::DateTime, looking_for_home::ListLookingForAdoptionVariables,
};
use snafu::{OptionExt, ResultExt};
use std::env;

use crate::{
    entity::{get_entity, list_entity},
    models::AdviceArticle,
    queries::{
        commons::{BooleanFilterInput, StringFilterInput},
        external_media::ListExternalMediaVariables,
        found_cat::ListFoundCatVariables,
        lost_cat::ListLostCatVariables,
    },
};
mod entity;

// this should work fine but breaks rust-analyzer
// pub type Result<T, E = Error> = std::result::Result<T, E>;
pub fn get_announcement_article(announcement_id: String) -> Result<Article, Error> {
    use queries::announcement_article::GetArticleVariables;
    let id: cynic::Id = announcement_id.into();
    let vars = GetArticleVariables { id };
    get_entity::<Article>(vars)
}
pub fn get_advice_article(advice_id: String) -> Result<Article, Error> {
    use queries::advice_article::GetAdviceArticleVariables;
    let id: cynic::Id = advice_id.into();
    let vars = GetAdviceArticleVariables { id };
    Ok(get_entity::<AdviceArticle>(vars)?.into())
}
pub fn list_external_media(
    options: Options<ExternalMediaFilter>,
) -> Result<Paged<ExternalMedia>, Error> {
    let pagination = options.pagination;
    let sort: Option<Vec<Option<String>>> = match options.sort {
        empty if empty.is_empty() => None,
        otherwise => Some(otherwise.into_iter().map(Some).collect()),
    };

    let vars = ListExternalMediaVariables {
        filters: None,
        pagination,
        sort,
    };
    list_entity::<ExternalMedia>(vars)
}
pub fn list_announcement(
    options: Options<AnnouncementFilter>,
) -> Result<Paged<Announcement>, Error> {
    use queries::announcement::ListAnnouncementsVariables;

    let pagination = options.pagination;
    let sort: Option<Vec<Option<String>>> = match options.sort {
        empty if empty.is_empty() => None,
        otherwise => Some(otherwise.into_iter().map(Some).collect()),
    };

    let vars = ListAnnouncementsVariables {
        filters: None,
        pagination,
        sort,
    };
    list_entity::<Announcement>(vars)
}
pub fn list_advice(options: Options<AdviceFilter>) -> Result<Paged<Advice>, Error> {
    use queries::advice::ListAdviceVariables;

    let pagination = options.pagination;
    let sort: Option<Vec<Option<String>>> = match options.sort {
        empty if empty.is_empty() => None,
        otherwise => Some(otherwise.into_iter().map(Some).collect()),
    };

    let vars = ListAdviceVariables {
        filters: None,
        pagination,
        sort,
    };
    list_entity::<Advice>(vars)
}

pub fn get_cat(id: String) -> Result<Cat, Error> {
    use queries::cat::GetCatVariables;

    let id: cynic::Id = id.into();
    get_entity::<Cat>(GetCatVariables { id })
}

pub fn list_adopted_cat(
    options: Options<CatFilter>,
    between_dates: Option<BetweenDateTime>,
) -> Result<Paged<AdoptedCat>, Error> {
    use crate::queries::adopted_cat::ListAdoptedCatVariables;

    let filters: CatFiltersInput = options.filter.map_or_else(
        || CatFiltersInput {
            is_dead: Some(BooleanFilterInput {
                eq: Some(false),
                ..BooleanFilterInput::default()
            }),
            ..CatFiltersInput::default()
        },
        |filter| CatFiltersInput {
            is_dead: Some(BooleanFilterInput {
                eq: Some(false),
                ..BooleanFilterInput::default()
            }),
            ..filter.into()
        },
    );
    let pagination = options.pagination.unwrap_or_default();
    let sort: Option<Vec<Option<String>>> = match options.sort {
        empty if empty.is_empty() => None,
        otherwise => Some(otherwise.into_iter().map(Some).collect()),
    };

    let between: Option<Vec<Option<DateTime>>> = between_dates.map(|dates| dates.into());
    let vars = ListAdoptedCatVariables {
        cat: filters,
        pagination,
        sort,
        between,
    };

    list_entity::<AdoptedCat>(vars)
}

impl CatFiltersInput<'_> {
    fn from_slug(slug: String) -> Self {
        CatFiltersInput {
            slug: Some(StringFilterInput {
                eq: Some(slug),
                ..Default::default()
            }),
            ..Default::default()
        }
    }
}
pub fn get_found_cat_by_slug(slug: String) -> Result<FoundCat, Error> {
    let key = serde_json::to_string(&slug);
    let filters = CatFiltersInput::from_slug(slug);
    let pagination = PaginationArg::default();

    let vars = ListFoundCatVariables {
        filters,
        pagination,
        sort: None,
    };
    let cats = list_entity::<FoundCat>(vars)?.items;
    assert!(cats.len() < 2);
    cats.into_iter()
        .next()
        .context(NotFoundSnafu { key: key.unwrap() })
}
pub fn get_lost_cat_by_slug(slug: String) -> Result<LostCat, Error> {
    let key = serde_json::to_string(&slug);
    let filters = CatFiltersInput::from_slug(slug);
    let pagination = PaginationArg::default();

    let vars = ListLostCatVariables {
        filters,
        pagination,
        sort: None,
    };
    let cats = list_entity::<LostCat>(vars)?.items;
    assert!(cats.len() < 2);
    cats.into_iter()
        .next()
        .context(NotFoundSnafu { key: key.unwrap() })
}

pub fn get_lfh_cat_by_slug(slug: String) -> Result<LookingForHomeCat, Error> {
    let key = serde_json::to_string(&slug);
    let filters = CatFiltersInput::from_slug(slug);
    let pagination = PaginationArg::default();

    let vars = ListLookingForAdoptionVariables {
        filters,
        pagination,
        owned_by_kotkowo: None,
        sort: None,
    };
    let cats = list_entity::<LookingForHomeCat>(vars)?.items;
    assert!(cats.len() < 2);
    cats.into_iter()
        .next()
        .context(NotFoundSnafu { key: key.unwrap() })
}

pub fn get_cat_by_slug(slug: String) -> Result<Cat, Error> {
    use crate::queries::cat::ListCatVariables;

    let key = serde_json::to_string(&slug);

    let filters: CatFiltersInput = CatFiltersInput {
        slug: Some(StringFilterInput {
            eq: Some(slug),
            ..Default::default()
        }),
        ..Default::default()
    };

    let pagination = PaginationArg::default();

    let vars = ListCatVariables {
        filters,
        pagination,
        sort: None,
    };

    let cats: Vec<Cat> = list_entity::<Cat>(vars)?.items;

    assert!(cats.len() < 2);
    let cat: Cat = cats
        .into_iter()
        .next()
        .context(NotFoundSnafu { key: key.unwrap() })?;

    Ok(cat)
}

pub fn list_supporters_with_virtual_cats(
    pagination: Option<PaginationArg>,
    sort: Option<Vec<String>>,
) -> Result<Paged<Supporter>, Error> {
    use queries::virtual_cat::ListSupporterWithCatsVariables;

    let pagination = pagination.unwrap_or_default();
    let sort: Option<Vec<Option<String>>> = sort.map(|s| s.into_iter().map(Some).collect());
    let vars = ListSupporterWithCatsVariables { pagination, sort };

    list_entity::<Supporter>(vars)
}

pub fn list_lost_cat(options: Options<CatFilter>) -> Result<Paged<LostCat>, Error> {
    use queries::lost_cat::ListLostCatVariables;
    let filters: CatFiltersInput = options
        .filter
        .map_or_else(CatFiltersInput::default, |filter| filter.into());
    let pagination = options.pagination.unwrap_or_default();
    let sort: Option<Vec<Option<String>>> = match options.sort {
        empty if empty.is_empty() => None,
        otherwise => Some(otherwise.into_iter().map(Some).collect()),
    };
    let vars = ListLostCatVariables {
        filters,
        pagination,
        sort,
    };
    list_entity::<LostCat>(vars)
}

pub fn list_found_cat(options: Options<CatFilter>) -> Result<Paged<FoundCat>, Error> {
    use queries::found_cat::ListFoundCatVariables;

    let filters: CatFiltersInput = options
        .filter
        .map_or_else(CatFiltersInput::default, |filter| filter.into());
    let pagination = options.pagination.unwrap_or_default();
    let sort: Option<Vec<Option<String>>> = match options.sort {
        empty if empty.is_empty() => None,
        otherwise => Some(otherwise.into_iter().map(Some).collect()),
    };
    let vars = ListFoundCatVariables {
        filters,
        pagination,
        sort,
    };

    list_entity::<FoundCat>(vars)
}

pub fn list_looking_for_adoption_cat(
    options: Options<CatFilter>,
    owned_by_kotkowo: Option<bool>,
) -> Result<Paged<LookingForHomeCat>, Error> {
    use queries::looking_for_home::ListLookingForAdoptionVariables;

    let filters: CatFiltersInput = options
        .filter
        .map_or_else(CatFiltersInput::default, |filter| filter.into());
    let pagination = options.pagination.unwrap_or_default();
    let sort: Option<Vec<Option<String>>> = match options.sort {
        empty if empty.is_empty() => None,
        otherwise => Some(otherwise.into_iter().map(Some).collect()),
    };
    let vars = ListLookingForAdoptionVariables {
        owned_by_kotkowo,
        filters,
        pagination,
        sort,
    };

    list_entity::<LookingForHomeCat>(vars)
}

pub fn list_virtual_cat(options: Options<CatFilter>) -> Result<Paged<Cat>, Error> {
    use crate::queries::virtual_cat::ListVirtualCatVariables;
    use cynic::http::ReqwestBlockingExt;
    use cynic::QueryBuilder;
    use queries::virtual_cat::ListVirtualCat;

    let endpoint = env::var("STRAPI_ENDPOINT").context(EnvVarMissingSnafu {})?;

    let filters: CatFiltersInput = options
        .filter
        .map_or_else(CatFiltersInput::default, |filter| filter.into());
    let pagination = options.pagination.unwrap_or_default();
    let sort: Option<Vec<Option<String>>> = match options.sort {
        empty if empty.is_empty() => None,
        otherwise => Some(otherwise.into_iter().map(Some).collect()),
    };
    let vars = ListVirtualCatVariables {
        filters,
        pagination,
        sort,
    };

    let vars_str = serde_json::to_string(&vars);

    let operation = ListVirtualCat::build(vars);
    let query = operation.query.clone();
    let client = get_client()?;
    let response = client
        .post(endpoint)
        .run_graphql(operation)
        .context(CynicRequestSnafu {})?;

    if let Some(err) = response.errors {
        let message = format!(
            "Variables:\n{}\nGraphQL:\n{}\nError:\n{:?}",
            vars_str.unwrap(),
            query,
            err
        )
        .to_string();

        return Err(Error::RequestResultedInError { message });
    }

    let source_cats = response
        .data
        .context(MissingAttributeSnafu {})?
        .virtual_cats
        .context(MissingAttributeSnafu {})?;

    let meta = source_cats.meta;

    let cats: Result<Vec<Cat>, Error> = source_cats
        .data
        .into_iter()
        .map(|cat_entity| {
            cat_entity
                .attributes
                .context(MissingAttributeSnafu {})
                .and_then(|virtual_cat| {
                    Ok(virtual_cat
                        .cat
                        .context(MissingAttributeSnafu {})?
                        .data
                        .context(MissingAttributeSnafu {})?
                        .attributes
                        .context(MissingAttributeSnafu {})?
                        .into())
                })
        })
        .collect();

    let page: Paged<Cat> = Paged::new(meta.pagination, cats?);

    Ok(page)
}

pub fn list_cat(options: Options<CatFilter>) -> Result<Paged<Cat>, Error> {
    use crate::queries::cat::ListCatVariables;

    let filters: CatFiltersInput = options
        .filter
        .map_or_else(CatFiltersInput::default, |filter| filter.into());
    let pagination = options.pagination.unwrap_or_default();
    let sort: Option<Vec<Option<String>>> = match options.sort {
        empty if empty.is_empty() => None,
        otherwise => Some(otherwise.into_iter().map(Some).collect()),
    };
    let vars = ListCatVariables {
        filters,
        pagination,
        sort,
    };

    list_entity::<Cat>(vars)
}

fn get_client() -> Result<reqwest::blocking::Client, Error> {
    let api_key = env::var("STRAPI_KEY").context(EnvVarMissingSnafu {})?;
    let mut headers = reqwest::header::HeaderMap::with_capacity(1);
    headers.insert(
        reqwest::header::AUTHORIZATION,
        format!("Bearer {}", api_key)
            .parse()
            .context(InvalidHeaderValueSnafu {})?,
    );
    let client = reqwest::blocking::Client::builder()
        .default_headers(headers)
        .build()
        .context(RequestSnafu {})?;

    Ok(client)
}

#[cfg(test)]
mod tests {
    use crate::{
        get_announcement_article, get_cat, get_cat_by_slug, list_adopted_cat, list_announcement,
        list_cat, list_external_media, list_found_cat, list_looking_for_adoption_cat,
        list_lost_cat, list_supporters_with_virtual_cats, list_virtual_cat, Options, PaginationArg,
    };

    #[test]
    fn list_announcement_test() {
        let paged = list_announcement(Options::default());
        assert!(paged.is_ok());
    }
    #[test]
    fn get_announcement_article_test() {
        let article = get_announcement_article("4".to_string());
        assert!(article.is_ok());
    }
    #[test]
    fn list_adopted_cat_test() {
        let paged = list_adopted_cat(Options::default(), None);
        assert!(paged.is_ok());
    }

    #[test]
    fn list_lost_cat_test() {
        let opts = Options::default();
        let paged = list_lost_cat(opts);
        assert!(paged.is_ok())
    }
    #[test]
    fn list_found_cat_test() {
        let opts = Options::default();
        let paged = list_found_cat(opts);
        assert!(paged.is_ok())
    }

    #[test]
    fn list_virtual_cat_test() {
        let opts = Options::default();
        let paged = list_virtual_cat(opts);
        assert!(paged.is_ok())
    }

    #[test]
    fn list_supporters_with_cats_test() {
        let pagination = PaginationArg {
            ..Default::default()
        };
        let paged = list_supporters_with_virtual_cats(Some(pagination), None);
        assert!(paged.is_ok())
    }
    #[test]
    fn list_cat_test() {
        let opts = Options::default();
        let paged = list_cat(opts);
        assert!(paged.is_ok())
    }
    #[test]
    fn list_lfh_cat_test() {
        let opts = Options::default();
        let paged = list_looking_for_adoption_cat(opts, None);
        assert!(paged.is_ok())
    }
    #[test]
    fn get_cat_test() {
        let slug = "1";
        let cat = get_cat(slug.to_string());
        assert!(cat.is_ok());
    }
    #[test]
    fn get_cat_by_slug_test() {
        let slug = "luna";
        let cat = get_cat_by_slug(slug.to_string());
        assert!(cat.is_ok())
    }
    #[test]
    fn get_external_media() {
        let opts = Options::default();
        let media = list_external_media(opts);
        assert!(media.is_ok());
    }
}
