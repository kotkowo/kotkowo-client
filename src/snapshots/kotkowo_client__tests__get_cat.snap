---
source: src/lib.rs
expression: "format!(\"{:?}\", resp)"
---
Ok(Cat { name: "Kassadin" })
